presto-main/src/main/java/com/facebook/presto/server/remotetask/HttpRemoteTask.java

```diff
@@ -402,28 +412,53 @@ private synchronized void updateSplitQueueSpace()
                 unmodifiableWhenSplitQueueHasSpace = unmodifiableFuture(whenSplitQueueHasSpace);
             }
         }
+        return toComplete;
     }
 
-    private synchronized void processTaskUpdate(TaskInfo newValue, List<TaskSource> sources)
+    private void fireSplitQueueHasSpace(@Nullable CompletableFuture<?> future)
     {
-        updateTaskInfo(newValue);
-
-        // remove acknowledged splits, which frees memory
-        for (TaskSource source : sources) {
-            PlanNodeId planNodeId = source.getPlanNodeId();
-            int removed = 0;
-            for (ScheduledSplit split : source.getSplits()) {
-                if (pendingSplits.remove(planNodeId, split)) {
-                    removed++;
-                }
+        checkState(!Thread.holdsLock(this), "Can not notify split queue future while holding the lock");
+
+        if (future == null) {
+            return;
+        }
+
+        executor.execute(() -> {
```

Comment:
This change moves the complete to another thread, so you don't need the synchronization changes to assure that you are not competing in a lock